kernel/time.c

```diff
@@ -122,3 +123,22 @@ dword_t sys_times(addr_t tbuf) {
     }
     return 0;
 }
+
+dword_t sys_gettimeofday(addr_t tv, addr_t tz) {
+    STRACE("gettimeofday(0x%x, 0x%x)", tv, tz);
+    struct timeval timeval;
+    struct timezone timezone;
+    if (gettimeofday(&timeval, &timezone) < 0) {
+	    return errno_map();
+    }
+    struct timeval_ tv_;
+    struct timezone_ tz_;
+    tv_.sec = timeval.tv_sec;
+    tv_.usec = timeval.tv_usec;
+    tz_.minuteswest = timezone.tz_minuteswest;
+    tz_.dsttime = timezone.tz_dsttime;
+    if ((tv && user_put(tv, tv_)) || (tz && user_put(tz, tz_))) {
+	    return _EFAULT;
+    }
```

Comment:
If tv is a valid pointer and tz is a nonzero invalid pointer, this will write out tv, when it should do nothing and return EFAULT. I don't think this is super important to fix though.

Reply:
I took a moment to try out a variety of other Linuxes (Linuces?) do in this case, and I got different results, none of which actually match what you've mentioned. For systems with glibc, the program will SIGSEGV because it will [try to dereference `tz`](https://github.com/bminor/glibc/blob/master/sysdeps/posix/gettimeofday.c#L51), on others, where it actually performs the syscall, [`tv` will be set](https://github.com/torvalds/linux/blob/71e56028173bc84f01456a5679d8be9d681b49f1/kernel/time/time.c#L144). Honestly, I think we can consider behavior to be unspecified in this case, since you shouldn't have any expectation on what happens if you pass in non-`NULL` pointers that you shouldn't be passing in. I don't think POSIX specifies we do anything in this case.

Reply:
If linux has the same behavior, then it's definitely fine.